pilot/pkg/xds/endpoint_builder.go

```diff
@@ -123,12 +147,93 @@ func (b *EndpointBuilder) canViewNetwork(network string) bool {
 	return b.networkView[network]
 }
 
+// TODO(lambdai): Receive port value(15009 by default), builder to cover wide cases.
+type EndpointTunnelApplier interface {
+	// Mutate LbEndpoint in place. Return non-nil on failure.
+	ApplyTunnel(lep *endpoint.LbEndpoint, tunnelType networking.TunnelType) (*endpoint.LbEndpoint, error)
+}
+
+type EndpointNoTunnelApplier struct{}
+
+// Note that this will not return error if another tunnel typs requested.
+func (t *EndpointNoTunnelApplier) ApplyTunnel(lep *endpoint.LbEndpoint, tunnelType networking.TunnelType) (*endpoint.LbEndpoint, error) {
+	return lep, nil
+}
+
+type EndpointH2TunnelApplier struct{}
+
+// TODO(lambdai): Set original port if the default cluster original port is not the same.
+func (t *EndpointH2TunnelApplier) ApplyTunnel(lep *endpoint.LbEndpoint, tunnelType networking.TunnelType) (*endpoint.LbEndpoint, error) {
+	switch tunnelType {
+	case networking.H2Tunnel:
+		if ep := lep.GetEndpoint(); ep != nil {
+			if port := ep.Address.GetSocketAddress().GetPortSpecifier(); port != nil {
```

Comment:
should maybe use `GetPortValue` to avoid cast and potential panic

Reply:
Good catch! Previous it was an in place port value change. Now that I copy on write and there is a simpler assignment at below